[OpenMP][OpenMPIRBuilder] Move copyInput to a passed in lambda function and re-order kernel argument load/stores

agozillon opened the pull request:
This patch moves the existing copyInput function
into a lambda argument that can be defined
by a caller to the function.

This allows more flexibility in how the function
is defined, allowing Clang and MLIR to utilise
their own respective functions and types inside
of the lamba without affecting the OMPIRBuilder
itself.

The idea is to eventually replace/build on
the existing copyInput function that's used
and moved into OpenMPToLLVMIRTranslation.cpp
to a slightly more complex implementation
that uses MLIRs map information (primarily
ByRef and ByCapture information at the
moment).

The patch also moves kernel load stores to the top
of the kernel, prior to the first openmp runtime 
invocation. Just makes the IR a little closer to Clang.

agozillon commented on llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:
```diff
@@ -2166,6 +2166,9 @@ class OpenMPIRBuilder {
       InsertPointTy AllocaIP, InsertPointTy CodeGenIP)>;
 
+  using TargetGenArgAccessorsCallbackTy = function_ref<Value *(
+      Argument &Arg, Value *Input, IRBuilderBase &Builder)>;
```
Yes, it appears to be necessary, the OpenMPIRBuilder has it's own LLVM builder that it utilises internally which will end up with a different InsertionPoint as the lowering of createTarget progresses, All we currently do is pass an initial insertion point to createTarget and then the internal OpenMPIRBuilder's LLVM builder takes this insertion point and progresses it and then returns the end insertion point back up to the original ModuleTranslation's LLVM builder so it can continue working from wherever the OpenMPIRBuilder left off! That's how I understand it at least. So using the local copy by reference will result in inserting the new instructions at the incorrect location as it won't ever actually be used by the OpenMPIRBuilder.

TIFitis replied:
I don't think passing around the builder is the coorect approach to solve this.

We should be passing and returning the inserpoints instead. We should pass the AllocaIP and CodeGenIP to the Callback, and the CB should return the new position of the IP.

If new allocas are not needed the we can perhaps ignore the AllocaIP.

agozillon replied:
I don't think the AllocaIP is currently used, the insertion point that's used at the moment is the first non-phi or debug node, which does result in slightly differing IR to Clang. But I'll utilise the insertion points pass in method and try to get it the IR closer to what Clang currently has.

My only worry of using two differing builders is if they have differing states, but I'm not sure if that is a concern, perhaps the builders only relevant state is the insertion points.

shraiysh commented on llvm/unittests/Frontend/OpenMPIRBuilderTest.cpp:
```diff
@@ -5301,6 +5319,23 @@ TEST_F(OpenMPIRBuilderTest, TargetRegionDevice) {
       Constant::getNullValue(PointerType::get(Ctx, 0))};
 
+  auto SimpleArgAccessorCB = [&](llvm::Argument &Arg, llvm::Value *Input,
+                                 IRBuilderBase &Builder) {
```
Should this callback also take an AllocaIP? Without that, we make it rely on the caller to make sure builder is set at the right IP.

agozillon replied:
It's currently only ever called by createOutlinedFunction, which does pre-set the insertion point (and currently doesn't use the AllocaIP which results in slightly differing IR to Clang currently) and takes a lot of ownership of where that insertion point is at the moment, so this mainly just keeps the status quo that exists right now. 

However, I am more than happy to allow it to take an IP or perhaps two, to make the IR from MLIR vs Clang look a little more identical and this would also allow me to utilise the ModuleTranslation builder as @TIFitis suggests

shraiysh replied:
I don't think differing IRs is a problem as long as they are correct. I have not worked much on target side, so I am okay with the current implementation and would also be okay with adding an IP, whatever works best.

agozillon commented:
Updated to utilise the InsertPoint style interface for the lambda, hopefully it's a little more inline with what is expected for the CallbackInterfaces! Please let me know if it's not and if there's any other changes you'd like made.

I also tried to generate output a little more like Clang, with the Allocas/Store preamble for the argument in the entry block and the loads in the user block. You can see what I mean a bit more tangibly with the altered tests. If we're not comfortable with that I can simply use the same InsertPoint for loads/allocas/stores as before. However, this change doesn't affect any runtime tests we have at the moment.

TIFitis reviewed (approved):
LGTM 😄

agozillon commented:
Thank you very much for the review guys! 

I'll give @jsjodin sometime to review it if he wishes as he wrote it originally, and incase @jdoerfert wishes to give any input as well. Neither are required if you do not wish to spend the time reviewing though (as I'm aware everyone's time is precious and people are likely preparing to go to the conference next week), but I'll leave the patch open until Friday afternoon/evening (UK time) just incase!

agozillon committed:
[OpenMP][OpenMPIRBuilder] Move copyInput to a passed in lambda function and re-order kernel argument load/stores

This patch moves the existing copyInput function
into a lambda argument that can be defined
by a caller to the function.

This allows more flexibility in how the function
is defined, allowing Clang and MLIR to utilise
their own respective functions and types inside
of the lamba without affecting the OMPIRBuilder
itself.

The idea is to eventually replace/build on
the existing copyInput function that's used
and moved into OpenMPToLLVMIRTranslation.cpp
to a slightly more complex implementation
that uses MLIRs map information (primarily
ByRef and ByCapture information at the
moment).

The patch also moves kernel load stores to the top
of the kernel, prior to the first openmp runtime
invocation. Just makes the IR a little closer to Clang.
